Bug fix/Make OAX weapons work with Jump

HeavensSword opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->

**_I affirm:_**

- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/AirSkyBoat/AirSkyBoat/blob/staging/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/AirSkyBoat/AirSkyBoat/blob/staging/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## Please enter a player-facing description

<!-- Example: Adjusted the damage limits on physical weaponskills (Shozokui) -->
* Fixed WS calculation for Jump type abilities to take "Occasionally Attacks X" weapons into account. (Both MainHand and OffHand) (HeavensSword)
* Fixed Jump TP calculation when multi-hits occur. (HeavensSword)
* Made it so Jump abilities will no longer give TP to the target. (This is how retail functions) (HeavensSword)

## What does this pull request do? (Please be technical)

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->
* WS can now utilize a new parameter "canUseOccAtkX". Currently this is only needed for Jump abilities, but I went with a more flexible approach in case a use case arises where a WS or Ability that utilizes WS DMG calcs like Jump need to respect OAX Weapons.
* I've exposed the GetHitCount method of CItemWeapon through the BattleEntity so that it can now be called from Lua. 
* Previously the Jump multi-hit TP calculation was incorrectly changing the tpMultiplier which could result in higher than expected values. I've changed it to instead collate the total hits into tpHitsLanded.
*  The last major change was adding a new argument to the takeWeaponskillDamage methods that is just a boolean for whether the WS was a Jump ability. This is necessary because Jumps do not actually give TP to defenders on retail, so a check has been added to avoid that happening.
* Translated a number of the old Russian code comments for clarity.

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here. -->
The easiest way to test these changes is to make your GM character a WAR/DRG and equip a Ridill. (Avoid using Store TP Gear)
You should then be able to use Jump and High Jump on a mob and observe your TP gain. I believe for Ridill's delay, the base gain from 1 hit is 63, but you can figure it out by either waiting for a single auto-attack, or switch to a Hofud and observe the TP gain to get your base.
You should see 63 TP * the number of OAX hits for that activation. 
EDIT: If DA/TA/QA proc, they take precedence over OAX calculation.

If you want some fun, switch to NIN/DRG and equip Kraken Club and Merc. Kris. This helps validate the behavior with Dual Wield.

## Special Deployment Considerations

<!-- Include any steps that need to be taken when deploying to the live environment. -->
<!-- Example: Need to run one_time_sql_conversion.sql -->

There shouldn't be any special considerations.

TracentEden commented on CMakeSettings.json:
Should remove this file form the PR by changing back to old version and then committing, then also should squash your commits into a single commit as will help when merging

HeavensSword replied:
Will do, thanks!

TracentEden commented on scripts/globals/weaponskills.lua:
```diff
@@ -250,7 +251,29 @@ local function getMultiAttacks(attacker, target, wsParams)
     end
 
-    numHits = utils.clamp(numHits + bonusHits, 1, 8)
+    -- If this WS/Ability is capable of multi-striking from Occasionally Attacks X weapons, calculate now
```
I think in most cases you cannot have both bonusHits and totalOAXHIts apply (see https://www.ffxiah.com/forum/topic/31534/math-oax-multi-attack-and-you/#1926857 and also the normal attack code for this) so if bonusHits is >0 then totalOAXHits should be 0.

HeavensSword replied:
Ah I think you are correct; that was my misunderstanding of ["Occasionally attacks X times" statistics are processed after [Quadruple Attack](https://www.bg-wiki.com/ffxi/Quadruple_Attack), [Triple Attack](https://www.bg-wiki.com/ffxi/Triple_Attack), and [Double Attack](https://www.bg-wiki.com/ffxi/Double_Attack) checks](https://www.bg-wiki.com/ffxi/Occasionally_Attacks_X_Times). I'll fix it.

TracentEden replied:
Can just be bonusHits == 0 I think?

HeavensSword replied:
Updated

HeavensSword commented:
Thanks for the help @TracentEden! I've squashed and merged in latest from Staging. I've also adjusted the OAX check to skip if bonusHits is > 0. 

Let me know if there are any other changes you need.

TracentEden commented on scripts/globals/weaponskills.lua:
```diff
@@ -250,7 +251,31 @@ local function getMultiAttacks(attacker, target, wsParams)
+    then
+        local numMainHits = attacker:getWeaponHitCount()
+        -- Subtract 1 since the initial hit would have been calculated already; we just want additionals
+        numMainHits = utils.clamp(numMainHits - 1, 0, numMainHits)
```
Do we need this clamp if numMainHits will always be at least 1 or can it be 0? Same for the offhand clamp

HeavensSword replied:
Just verified; the value ultimately comes from battleutils.cpp "getHitCount" which is indeed minimum of 1 and already caps at 8 in the return.

I'll remove the redundant clamps.

HeavensSword commented:
I've updated my commit message to reflect the change with regards to DA/TA/QA procs.

TracentEden commented on src/map/entities/battleentity.cpp:
```diff
@@ -546,9 +546,29 @@ uint16 CBattleEntity::GetRangedWeaponRank()
+
 /************************************************************************
  *                                                                       *
- *  Изменяем количество TP сущности                                      *
```
I think you can remove from the PR the changes to comments as when we rebase ASB to LSB then I think they already have changed these comments there. Thanks!

HeavensSword replied:
Gotcha, I've reverted the translations

TracentEden replied:
> Gotcha, I've reverted the translations

I still see a few comment changes leftover

HeavensSword replied:
Oh whoops I see; would you like me to revert the few I capitalized as well?

HeavensSword replied:
Reverted the one I missed and also the ones I capitalized where I also did not change any code.

TracentEden replied:
Thanks, would you be able to check this [PR](https://github.com/LandSandBoat/server/pull/4783/files) as well and align your PR as much as possible with regard to things such as function, variable names, and code (without like creating new files or such). This will help later on when we merge in LSB. Sorry for the further request, a lot of moving parts on these kind of changes.

HeavensSword replied:
I notice that LSB is using from different conventions in code than ASB. e.g. I noticed other areas of the code in ASB using assertions such as : `XI_DEBUG_BREAK_IF(m_PBaseEntity->objtype == TYPE_NPC);` while in LSB they seem to print a warning instead.
Is the rule of thumb to follow LSB? Is that the more "up to date" code base so to speak?

TracentEden replied:
> I notice that LSB is using some different conventions in code than ASB. e.g. I noticed other areas of the code in ASB using assertions such as : `XI_DEBUG_BREAK_IF(m_PBaseEntity->objtype == TYPE_NPC);` while in LSB they seem to print a warning instead. Is the rule of thumb to follow LSB? Is that the more "up to date" code base so to speak?

yeah use LSB code in such cases

HeavensSword commented:
- I've merged my getWeaponHitCount methods into 1 using a bool param for offHand similar to what WinterSolstice did.
- I've also borrowed his solution for not feeding a Jump's target TP by setting TargetTPMult to 0. This simplifies some methods so I don't have to forward an extra argument anymore.
- I also fixed a potential issue I caught where I was returning 1 instead of 0 if a weapon was null.

TracentEden commented:
> * I've merged my getWeaponHitCount methods into 1 using a bool param for offHand similar to what WinterSolstice did.
> * I've also borrowed his solution for not feeding a Jump's target TP by setting TargetTPMult to 0. This simplifies some methods so I don't have to forward an extra argument anymore.
> * I also fixed a potential issue I caught where I was returning 1 instead of 0 if a weapon was null.

Yeah but you are still using like different code, take their code like in getWeaponHitCount and use directly, and same for other code as long as it does not create new files or such

Spaceballs-the-galka commented:
Noticed this PR has been stuck for a while, unsure if HeavenSword lost steam or if he is dealing with that nasty thing called life.

What work needs to be carried forward here to get this to completion? @TracentEden

Spaceballs-the-galka commented:
PS thats me wanting to help, not be a dick and press yall

TracentEden commented:
> PS thats me wanting to help, not be a dick and press yall
> 

Sorry about the late reply, from my understanding this will for the most part be automatically fixed when we rebase to LSB (as zach has this fixed there).

Frankie-hz commented:
This will be fixed with the rebase as upstream has fixed it. Closing.
